pennylane/devices/qubit/sampling.py

```diff
@@ -28,13 +28,83 @@
 from .apply_operation import apply_operation
 
 
+def _group_measurements(mps: List[Union[SampleMeasurement, ClassicalShadowMP, ShadowExpvalMP]]):
+    """
+    Group the measurements such that:
+      - measurements with pauli observables pairwise-commute in each group
+      - measurements with observables that are not pauli words are all in different groups
+      - measurements without observables are all in the same group
+      - classical shadow measurements are all in different groups
+    """
+    if len(mps) == 1:
+        return [mps], [[0]]
+
+    mp_pauli_obs = []
+    mp_other_obs = []
+    mp_no_obs = []
+
+    for i, mp in enumerate(mps):
+        if isinstance(mp, (ClassicalShadowMP, ShadowExpvalMP)):
+            mp_other_obs.append((i, mp))
+        elif mp.obs is None:
+            mp_no_obs.append((i, mp))
+        elif isinstance(mp.obs, (Sum, Hamiltonian, SProd)):
+            # Sums and Hamiltonians are treated as valid Pauli words, but
+            # aren't accepted in qml.pauli.group_observables
+            mp_other_obs.append((i, mp))
+        elif qml.pauli.is_pauli_word(mp.obs):
+            mp_pauli_obs.append((i, mp))
+        else:
+            mp_other_obs.append((i, mp))
+
+    if mp_pauli_obs:
+        i_to_pauli_mp = dict(mp_pauli_obs)
+        ob_groups, group_indices = qml.pauli.group_observables(
+            [mp.obs for mp in i_to_pauli_mp.values()], list(i_to_pauli_mp.keys())
+        )
+
+        mp_pauli_groups = []
+        for group, indices in zip(ob_groups, group_indices):
+            mp_group = []
+            for ob, i in zip(group, indices):
+                mp_group.append(i_to_pauli_mp[i].__class__(obs=ob))
+
+            mp_pauli_groups.append(mp_group)
+    else:
+        mp_pauli_groups, group_indices = [], []
+
+    mp_no_obs_indices = [[mp[0] for mp in mp_no_obs]] if mp_no_obs else []
+    mp_no_obs = [[mp[1] for mp in mp_no_obs]] if mp_no_obs else []
```

Comment:
I think I'd be on board with directly populating these in the above for loop, instead of populating the `mp_no_obs` with a tuple and then splitting it up.